HW2/P4/driver.py

```diff
@@ -17,15 +17,67 @@
 
 def py_median_3x3(image, iterations=10, num_threads=1):
     ''' repeatedly filter with a 3x3 median '''
+    # create a new memory location with image data for tmpA
     tmpA = image.copy()
+    # create an empty structure for tmpB with same dimensions as tmpA
     tmpB = np.empty_like(tmpA)
+    # initialize num_threads-sized list for threads
+    threads = [None] * num_threads
+    # initialize (num-threads * (iterations)-sized list for events
+    events = [None] * num_threads * (iterations)
+
+    # initialize events across all elements in list
+    for threadidx in range(num_threads * (iterations)):
+        # initialize each position with an event
+        events[threadidx] = threading.Event()
+    
+    # reshape the events list to correspond to a 2D matrix
+    events = np.reshape(events, [num_threads, iterations])
+
+    # iterate for each thread
+    for threadidx in range(num_threads):
+        # initialize each thread with the appropriate arguments to the worker function
+        threads[threadidx] = threading.Thread(target = worker, args = (threadidx, events, tmpA, tmpB, num_threads, iterations))
+        # exit even when this thread is alive
+        threads[threadidx].daemon = True
+        # start the thread
+        threads[threadidx].start()
+
+    # collect and end the threads
+    for t in threads:
+        t.join()
+
+    # return our result
+    return tmpA
+
 
+def worker(threadidx, events, tmpA, tmpB, num_threads, iterations):
+    # initialize the list of adjacent threads
+    adj_threads = []
+    # set up adjacent threads (representing n-1 and n+1 threads) 
+    for i in range(threadidx-1, threadidx+2):
+        # exclude indices corresponding to edges
+        if (i >= 0 and i < num_threads):
+            adj_threads.append(i)
+            
+    # 10 interations
     for i in range(iterations):
-        filtering.median_3x3(tmpA, tmpB, 0, 1)
-        # swap direction of filtering
+        # skip event-checking if this is the first iteration
+        if i != 0:
+            # iterate only among adjacent threads (e.g., other threads are not touched)
+            for j in adj_threads:
+                # for these threads, block until the filtering/switching has occurred
+                events[j, i-1].wait()
+
+        # call median_3x3 Cython code with threadidx-multiple rows and offset num_threads
+        filtering.median_3x3(tmpA, tmpB, threadidx, num_threads)
+        # switch only the pertinent values of tmpA and tmpB
```

Comment:
Thank you for including thorough comments, it made this section very easy to understand.